### DISCUSSION TITLE
gum spin with python functions

### DISCUSSION THREAD

**Discussion by dsrosen6** [charmbracelet/gum#547] [Q&A]
Hi everyone, I am writing a TUI/CLI that will utilize gum- it is python based and I’m using the example from the GitHub page (gum.py) as a template for the function I’m using for gum commands.

In some specific instances, I want to use gum spin but I want the command it runs/waits for to be another python function. But my understanding is I can only run gum spin with another shell based command via subprocess.

Is there something I’m missing, or is there perhaps a workaround I can use to essentially start an “indefinite” spinner that I can start, and then stop, at certain points in my script?

**Comment by wellkilo** [comment]
You are not missing a Python hook: gum spin launches an **external process**. It cannot directly call a Python function that exists in the current interpreter, and it does not expose a separate “start spinner / stop spinner” process-control API.

A clean workaround is to make the function a child mode of your Python program and let gum run that mode:

    import subprocess
    import sys

    def expensive_work():
        # your real function
        ...

    if len(sys.argv) > 1 and sys.argv[1] == "_expensive_work":
        expensive_work()
        raise SystemExit(0)

    subprocess.run(
        ["gum", "spin", "--title", "Working...", "--",
         sys.executable, __file__, "_expensive_work"],
        check=True,
    )

This preserves the useful behavior of gum spin: it stops exactly when the child exits and propagates the child's non-zero exit status. Put the reusable business logic in a normal function and keep the command-line dispatch thin.

If the function must stay in the same Python process (because it holds non-serializable state, open transactions, etc.), use a Python spinner/library or render a spinner thread yourself. Starting a dummy gum spin process and killing it later is fragile and can leave terminal state or child processes behind.

The current implementation confirms the boundary: [commandStart](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/spin/spin.go#L50-L96) constructs an exec.Cmd from the arguments and exits the spinner when that process finishes.
